src/Microsoft.TemplateEngine.Cli/New3Command.cs

```diff
@@ -281,37 +312,56 @@ private void HandlePostActions(TemplateCreationResult creationResult)
             postActionDispatcher.Process();
         }
 
-        private void DisplayTemplateList(bool showAll = false)
+        // Checks the result of TemplatesToDisplayInfoAbout()
+        // If they all have the same group identity, return them.
+        // Otherwise retun an empty list.
+        private IEnumerable<ITemplateInfo> TemplatesToShowDetailedHelpAbout()
         {
-            IReadOnlyCollection<ITemplateInfo> results = null;
+            IReadOnlyList<ITemplateInfo> candidateTemplates = TemplatesToDisplayInfoAbout();
 
-            try
+            if (candidateTemplates.Count() == 0)
             {
-                if (showAll)
-                {
-                    results = PerformAllTemplatesInContextQueryAsync().Result.Where(x => x.IsMatch).Select(x => x.Info).ToList();
+                return candidateTemplates;
+            }
 
-                    if(results.Count == 0)
-                    {
-                        results = PerformAllTemplatesInContextQueryAsync().Result.Select(x => x.Info).ToList();
-                    }
-                }
-                else
-                {
-                    results = _matchedTemplates.Where(x => x.IsMatch).Select(x => x.Info).ToList();
+            string firstGroupIdentity = candidateTemplates.First().GroupIdentity;
 
-                    if (results.Count == 0)
-                    {
-                        results = _matchedTemplates.Select(x => x.Info).ToList();
-                    }
-                }
+            if (candidateTemplates.Any(x => x.GroupIdentity != firstGroupIdentity))
```

Comment:
This would be a handy extension method (for Utils) - something like 

``` C#
public static bool AllAreTheSame<T, TValue>(this IEnumerable<T> items, Func<T, TValue> selector)
    where TValue : IEquatable<TValue>
{
    return items.AllAreTheSame(selector, (x, y) => x?.Equals(y) ?? (y == null));
}

public static bool AllAreTheSame<T, TValue>(this IEnumerable<T> items, Func<T, TValue> selector, IEqualityComparer<TValue> comparer = null)
{
    if (comparer == null)
    {
        comparer = EqualityComparer<TValue>.Default;
    }

    return items.AllAreTheSame(selector, comparer.Equals);
}

public static bool AllAreTheSame<T, TValue>(this IEnumerable<T> items, Func<T, TValue> selector, Func<TValue, TValue, bool> comparer = null)
{
    if (comparer == null)
    {
        comparer = EqualityComparer<T>.Default.Equals;
    }

    using (IEnumerator<T> enumerator  = items.GetEnumerator())
    {
        if (!enumerator.MoveNext())
        {
            return true; //If there are no elements they're all the same - maybe?
        }

        TValue value = selector(enumerator.Current);

        while (enumerator.MoveNext())
        {
            if (!comparer.Equals(value, enumerator.Current))
            {
                return false;
            }
        }

        return true;
    }
}
```